suse-edge/edge-image-builder
# EIB RPM air-gapped and non-air gapped installation implementation

What does this PR do?
---
Introduces a way to install RPMs/pacakges and their dependencies in an air-gapped environment while keeping the existing single/standalone RPM installation.

How is the RPM/package resolution handled?
---
We create an additional container inside of the EIB container. This container uses an image that is based on the ISO/Raw image that has been provided by the user. Inside this new container we download all dependencies for the specified RPMs/packages and ship them in a directory which is then copied out of the container and to the EIB combustion directory. From there this directory is converted to an RPM repository which is then used during the combustion phase to provide the needed RPMs so that the specified resources can be installed. A detailed explanation of the process + a diagram will come in a later PR.

How has this PR been tested?
---
The following functionality has been tested:

1. Install packages and dependencies by only providing a list of packages and a scc.suse.com registration code
2. Install a standalone RPM package
3. Install packages specified in a list of packages using a third-party package repository to install the specified packages and resolve their dependencies
4. Install RPMs (that are copied in the 'rpms' directory) using a third-party package repository to resolve any RPM dependencies
5. Full test including all of the above points in a single configuration

The aforementioned functionality was tested on the following images:

1. SLE-Micro.x86_64-5.5.0-Default-SelfInstall-GM.install.iso
2. SLE-Micro.x86_64-5.5.0-Default-RT-GM.raw

Environments where the EIB was tested:

1. SLES 15-SP5
2. Ubuntu 22.04

Notes on using the new functionality
---
**_Functionality has only been tested on Linux systems. Building on Mac currently has some issues!_**

In order to have a podman container inside of a podman container EIB needs to be executed in the following way:

1. On systems that run a podman `cgroupVersion: v1` the EIB container image needs to be built as `root` and also you need to pass the `--privileged` to the `podman run` command:
```bash
sudo podman build -t eib:dev .
```
```bash
sudo podman run --rm --privileged  ...
```
3.  On systems that run a podman `cgroupVersion: v2` you just need to pass the `--privileged` flag to the `podman run`command. Nothing else around the build process changes.

```bash
podman run --rm --privileged  ...
```

I am currently investigating if there is a possible workaround to (1).
To check your podman cgroup version you can do:

```bash
podman info | grep cgroupVersion
```

What comes after this PR?
---

1. [Additional repository](https://github.com/suse-edge/edge-image-builder/blob/main/pkg/image/definition.go#L68) enhancement + `gpg-check` configuration support
2. Unit tests for the repo package
3. Use and functionality documentation for the implementation
4. Find a possible fix for building on Mac
